### ISSUE TITLE
reward_mean converges, but kl is getting bigger and bigger

### ISSUE THREAD

**Issue by Mryangkaitong** [lvwerra/trl#204]
I use trl for ppo training on my model (a seq2seq generation model). The PPOConfig is as follows.

` config = PPOConfig(
          model_name="/search/ai/kaitongyang/RLHF_DEBUG/PPO_trl/small_glm",
          learning_rate=1e-5,
           batch_size=32,
           ppo_epochs=3,
           log_with="wandb",
           init_kl_coef=0.3,
           remove_unused_columns=False,
           mini_batch_size=8
)`

 At the end of the training, I found that no matter what query is input, the output is the same response. It seems that there is a problem with the kl divergence. It stands to reason Say I understand that a good ppo convergence is that the reward_mean is getting bigger and bigger, and then tends to be stable. The kl divergence starts to change, but it tends to be problematic later, but the kl divergence here is getting bigger and bigger, which leads to the model not learning well.

<img width="1196" alt="企业微信截图_6dff19a9-ffb4-49dd-b779-5c07699bcb9f" src="https://user-images.githubusercontent.com/23132307/223903246-6bb96eaa-c903-43a1-bd0d-1df9c6c2fd7c.png">
<img width="1602" alt="企业微信截图_24b8d208-41d9-40f1-8f7b-03a642fa42ec" src="https://user-images.githubusercontent.com/23132307/223903351-bfb89334-2502-4c6a-8d5e-c72dcf66c979.png">
<img width="1620" alt="企业微信截图_61ca246b-2b3c-4578-82e9-02a965db42df" src="https://user-images.githubusercontent.com/23132307/223903461-fade99c2-b2e4-4614-b841-7a83fa4ef9ba.png">
<img width="1596" alt="企业微信截图_872e6da9-961c-4485-a1d7-dac6dc0f10c0" src="https://user-images.githubusercontent.com/23132307/223903542-00c2c029-1a8b-4635-831d-96759aad4e7d.png">

What could be the reason for this, thank you very much

**Comment by Mryangkaitong** [comment]
> I have a similar question that why there is a peak in "loss/total" and "loss/policy"？
Is your kl becoming more and more negative? According to my understanding, the method of objective/kl is（https://github.com/lvwerra/trl/blob/main/trl/trainer/ppo_trainer.py#L848）(data["logprobs"] - data["ref_logprobs"]), if it becomes more and more negative, it means that data["logprobs"] is getting smaller and smaller, and data[" logprobs"] is the logits after softmax_log, that is, the logits are getting smaller and smaller? More and more unbelievable?

**Comment by Alexixu** [comment]
@Mryangkaitong No, the KL value becoming bigger and more positive. 
The policy changes the origin token distribution so the KL value increases.

**Comment by lvwerra** [comment]
Note that the KL-divergence (`objective-kl`) should never be negative. This can happen if you generate text with settings that are not pure sampling (e.g. early stopping strategies, minimum generation length etc.). Have you tried the generation settings used in the T5 example.

**Comment by Mryangkaitong** [comment]
Thanks a lot, it worked

**Comment by Mryangkaitong** [comment]
Hello author, it is possible to converge now, but when the convergence reaches a certain level, the reward_mean will decrease again, as shown in the figure below (roughly 70 steps is an epoch), is this normal?

**Comment by Mryangkaitong** [comment]
> Note that the KL-divergence (`objective-kl`) should never be negative. This can happen if you generate text with settings that are not pure sampling (e.g. early stopping strategies, minimum generation length etc.). Have you tried the generation settings used in the T5 example.

Hello author, it is possible to converge now, but when the convergence reaches a certain level, the reward_mean will decrease again, as shown in the figure below (roughly 70 steps is an epoch), is this normal?
<img width="1071" alt="截屏2023-03-12 上午11 58 36" src="https://user-images.githubusercontent.com/23132307/224523521-96f13582-a3c1-4579-993c-3e1fd897e55a.png">
<img width="1485" alt="截屏2023-03-12 上午11 58 43" src="https://user-images.githubusercontent.com/23132307/224523524-4533c2e0-cbe1-4986-996d-04143a5680f8.png">

**Comment by PanchenkoYehor** [comment]
> > Note that the KL-divergence (`objective-kl`) should never be negative. This can happen if you generate text with settings that are not pure sampling (e.g. early stopping strategies, minimum generation length etc.). Have you tried the generation settings used in the T5 example.
> 
> Hello author, it is possible to converge now, but when the convergence reaches a certain level, the reward_mean will decrease again, as shown in the figure below (roughly 70 steps is an epoch), is this normal?
> 
> `learning_rate=1e-5, batch_size=32, ppo_epochs=3, init_kl_coef=0.1, mini_batch_size=32`
> 

Hi @Mryangkaitong , have you managed to resolve negative kl issue by changing generation kwargs? What generation kwargs do you use for now?

**Comment by lvwerra** [comment]
Maybe saving the best model would be a good idea? At each step compare average reward and save if it is the best so far. Maybe also a learning rate scheduler could help.

**Comment by Mryangkaitong** [comment]
> Note that the KL-divergence (`objective-kl`) should never be negative. This can happen if you generate text with settings that are not pure sampling (e.g. early stopping strategies, minimum generation length etc.). Have you tried the generation settings used in the T5 example.

when I use generation settings
`max_length=512`
objective-kl become negative, but no settings max_length, my response become very very short

**Comment by Mryangkaitong** [comment]
> > > Note that the KL-divergence (`objective-kl`) should never be negative. This can happen if you generate text with settings that are not pure sampling (e.g. early stopping strategies, minimum generation length etc.). Have you tried the generation settings used in the T5 example.
> > 
> > 
> > Hello author, it is possible to converge now, but when the convergence reaches a certain level, the reward_mean will decrease again, as shown in the figure below (roughly 70 steps is an epoch), is this normal?
> > `learning_rate=1e-5, batch_size=32, ppo_epochs=3, init_kl_coef=0.1, mini_batch_size=32`
> 
> Hi @Mryangkaitong , have you managed to resolve negative kl issue by changing generation kwargs? What generation kwargs do you use for now?

https://github.com/lvwerra/trl/blob/main/examples/sentiment/scripts/t5-sentiment.py#L90

**Comment by lvwerra** [comment]
With the generation kwargs you linked above + `max_new_tokens` the model should always generate as many tokens as you want without negative KL divergence.
